contracts/vending/VendingMachine.sol

```diff
@@ -0,0 +1,169 @@
+/// @notice Contract implements a special update protocol to enable KEEP/NU
+///         token holders to wrap their tokens and obtain T tokens according
+///         to a fixed ratio. This will go on indefinitely and enable NU and
+///         KEEP token holders to join T network without needing to buy or
+///         sell any assets. Logistically, anyone holding NU or KEEP can wrap
+///         those assets in order to receive T. They can also unwrap T in
+///         order to go back to the underlying asset. There is a separate
+///         instance of this contract deployed for KEEP holders and a separate
+///         instance of this contract deployed for NU holders.
+contract VendingMachine is Ownable, IReceiveApproval {
+    using SafeERC20 for IERC20;
+    using SafeERC20 for T;
+
+    /// @notice Divisor for precision purposes, used to represent fractions.
+    uint256 public constant FLOATING_POINT_DIVISOR = 1e18;
+
+    /// @notice The token being wrapped to T (KEEP/NU).
+    IERC20 public immutable wrappedToken;
+
+    /// @notice T token contract.
+    T public immutable tToken;
+
+    /// @notice The ratio with which T token is converted based on the provided
+    ///         token being wrapped (KEEP/NU), expressed in 1e18 precision.
+    ///
+    ///         When wrapping:
+    ///           x [T] = amount [KEEP/NU] * ratio / FLOATING_POINT_DIVISOR
+    ///
+    ///         When unwrapping:
+    ///           x [KEEP/NU] = amount [T] * FLOATING_POINT_DIVISOR / ratio
+    uint256 public immutable ratio;
+
+    /// @notice The total balance of wrapped tokens for the given holder
+    ///         account. Only holders that have previously wrapped KEEP/NU to T
+    ///         can unwrap, up to the amount previously wrapped.
+    mapping(address => uint256) public wrappedBalance;
+
+    event Wrapped(
+        address indexed recipient,
+        uint256 wrappedTokenAmount,
+        uint256 tTokenAmount
+    );
+    event Unwrapped(
+        address indexed recipient,
+        uint256 tTokenAmount,
+        uint256 wrappedTokenAmount
+    );
+
+    /// @dev Sets the reference to `wrappedToken` and `tToken`. Initializes
+    ///      conversion ratio between wrapped token and T based on the provided
+    ///      `_tTokenAllocation` and `_wrappedTokenSupply`.
+    /// @param _wrappedToken Address to ERC20 token that will be wrapped to T
+    /// @param _tToken Address of T token
+    /// @param _wrappedTokenSupply The total supply of the token that will be
+    ///       wrapped to T
+    /// @param _tTokenAllocation The allocation of T this instance of Vending
+    ///        Machine will receive
+    constructor(
+        IERC20 _wrappedToken,
+        T _tToken,
+        uint256 _wrappedTokenSupply,
+        uint256 _tTokenAllocation
+    ) {
+        wrappedToken = _wrappedToken;
+        tToken = _tToken;
+        ratio =
+            (FLOATING_POINT_DIVISOR * _tTokenAllocation) /
+            _wrappedTokenSupply;
```

Comment:
What do you think about adding a few `required` for the input values? I know we control these parameters, but it costs nothing in terms of gas / code size.